### ISSUE TITLE
Bad defaults in skbuild.setuptools_wrap.parse_args

### ISSUE THREAD

**Issue by yonip** [scikit-build/scikit-build#345]
While working on #342 , I noticed a (somewhat) deeper issue with how arguments are parsed.
Currently, the way `parse_args` is set up in `setuptools_wrap.py`, the only way to set the build type is by setting it on the command line in the `dutils` section of the arguments (being the first one before any `--`), and if not set in that place specifically, both cmake and make will be set with `Release` as their build type. I am not sure of the scope of this issue, but is this intended behavior?
That is, should the first set of command line arguments override: 
1. Command line arguments passed to cmake explicitly? Should the default do so?
2. Command line arguments passed to make explicity? Should the default do so?
3. Arguments passed to setup via kwargs? Should the default do so?

I believe that the answers to these questions are to all except the first half of 3 to all be "no," but I just want to make sure the maintainers of this project agree.
Any way to fix this will change the design of the argument parsing functions non-trivially, so I would like to know what way would be best to solve this.
Since generally the latter command line arguments override the former, I suggest the following parsing psuedo-code: 
```lang:python
def parse_args(user={dutils:[...], cmake:[...], make:[...]}):
    hardcoded = {
        dutils: [...],
        cmake: [...],
        make: [...]
    }
    cli = {
        dutils: sys.argv.split('--', 2)[0],
        cmake: sys.argv.split('--', 2)[1],
        make: sys.argv.split('--', 2)[2]
    }

    dutils = hardcoded.dutil + user.dutil + cli.dutil
    cmake = hardcoded.dutil.get_cmake() + hardcoded.cmake + 
              user.dutil.get_cmake() + user.cmake + 
              cli.dutil.get_cmake() + cli.cmake
    make = hardcoded.dutil.get_make() + hardcoded.cmake.get_make() + hardcoded.make + 
             user.dutil.get_make() + user.cmake.get_make() + user.make + 
             cli.dutil.get_make() + cli.cmake.get_make() + cli.make
```
In essence, first check `cli`, then `user`, then `hardcoded`, and within each, look in `make`, then `cmake`, then `dutils`, if applicable.
This will probably work best with a complete rewrite of `parse_args` and removal/replacement of its helper functions, like `parse_skbuild_args` and `create_skbuild_argparser`.

I hope this writeup makes sense.

**Comment by jcfr** [comment]
> Currently, the way parse_args is set up in setuptools_wrap.py, the only way to set the build type is by 
> setting it on the command line in the dutils section of the arguments (being the first one before any --), 

Indeed, this corresponds to the `--build-type` argument added by scikit-build to the setup.py global options. 

### Single-config vs multi-config generator

> and if not set in that place specifically, both cmake and make will be set with Release as their build 
> type. I am not sure of the scope of this issue, but is this intended behavior?

Defaulting to `Release` is indeed the expected behavior.

That said the current implementation unconditionally specify `-DCMAKE_BUILD_TYPE:STRING=` at configure time and also `--config` when invoking cmake to build the project using the syntax `cmake --build /build/dir --config <config>`, it should be one or the other base on the type of generator.

In that regards, there are roughly two types of CMake generators:
* multi-config: Visual Studio, ....
* single-config:  Ninja, Make, Nmake, ... 

The key difference is that for multi-config generator, the build type is only known at built time whereas for single configuration generator it is a configure time parameter.

### Precedence of arguments

Arguments can be set in few different ways:
* command line
* keyword value passed to the setup function:
  * directly in `setup.py`
  * using `setup.cfg` (this is currently not tested)
* default values hard-coded in `setuptools_wrap.setup`

(In the future, we should also discuss how such value could be set in `pyproject.toml`)

Generally speaking value on the left take precedence on the one on the right : `command line` | `kewords` | `defaults`

This means that if a keyword is passed as `setup(..., foo=1)`, passing the command line argument `--foo 2` will lead to `foo` having the value `2`.

### The case of command line arguments

#### Description of the "type" of arguments

Now, as you described above, argument targeting different tools can be passed through the command line.

Generally speaking, within scikit-build we have:

```
/path/to/build-frontend  [build-frontend-args, ...]  [-- [build-backend-configure-args, ...] [-- [build-backend-buildtool-args]]]
```

which translate to this in the context of `setup.py`:

```
setup.py [setup-args|skbuild-args, ...] [-- [cmake-args, ...] [-- [build-tool-args, ...]]]
```

or more specifically (copied from [here](https://scikit-build.readthedocs.io/en/latest/usage.html#command-line-options)):

```
setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...] [skbuild_opts] [-- [cmake_opts] [-- [build_tool_opts]]]
```


#### The case of `--config` argument

In the current implementation of scikit-build, `--config` is a `skbuild_opts`

After being parsed by [create_skbuild_argparser](https://github.com/scikit-build/scikit-build/blob/3c6c2d162bf1182b8802fb2574e176427c893db2/skbuild/setuptools_wrap.py#L51-L68), the following happen with the corresponding value:

(1) Update of `cmake_args`
https://github.com/scikit-build/scikit-build/blob/3c6c2d162bf1182b8802fb2574e176427c893db2/skbuild/setuptools_wrap.py#L81

(2) Update of `build_tool_args`
https://github.com/scikit-build/scikit-build/blob/3c6c2d162bf1182b8802fb2574e176427c893db2/skbuild/setuptools_wrap.py#L86

(3) calling of `CMaker.make` with `make_args` (which is the same as `build_tool_args` and should be renamed for consistency)

https://github.com/scikit-build/scikit-build/blob/3c6c2d162bf1182b8802fb2574e176427c893db2/skbuild/setuptools_wrap.py#L537


(4) removal of `--config` from the final list of build tool arguments


https://www.python.org/dev/peps/pep-0517/#config-settings

That said, it is worth noting that [PEP 517](https://www.python.org/dev/peps/pep-0517/#id4) doesn't support this model.

Currently, since we use

**Comment by jcfr** [comment]
In my previous note, I described some terminology as well as details of the existing implementation.

### Back to the original question

Now I will try to answer your question (or at least clarify the terms)

> Should the first set of command line arguments override
>    Command line arguments passed to cmake explicitly? Should the default do so?
>    Command line arguments passed to make explicity? Should the default do so?
>    Arguments passed to setup via kwargs? Should the default do so?

Let's clarify further so that we talk about the same arguments:

* cmake command line arguments corresponds to `cmake_args`: They are passed to cmake for configuring the project  (e.g `-DFOO:STRING=bar`)

* make command line arguments corresponds to `build_tool_args`:
  * They are passed to the build tool like  `make`, `ninja`, or `msbuild`, an example of such parameter is `-jN` for `make` or `/m` for `msbuild`. 

* argument passed via `kwargs`. i am assuming we are talking about `cmake_args` here.


### An example

```
./setup.py bdist_wheel --config RelWithDebInfo -- -DCMAKE_BUILD_TYPE:STRING=Release
```

I think in this case, we should error out and clearly indicate there are contradictory  arguments.

**Comment by jcfr** [comment]
## Next steps

> This will probably work best with a complete rewrite of parse_args and removal/replacement of its helper functions, like parse_skbuild_args and create_skbuild_argparser.

Improving the approach for handling argument would be nice. But I anticipate it would require a fair amount of work.

In the mean time, may be the handling of `--build` argument could be improved so that `CMAKE_BUILD_TYPE` is only set for single-config generator.

And to answer the original comment:

> the only way to set the build type is by setting it on the command line

To support this, I suggest we introduce a new keyword named `cmake_build_type` that would be used to set `-DCMAKE_BUILD_TYPE:STRING=..` for single-config generator or used as the value for `--config ` for multi-config generator.



### Medium/Long term

We could look at 
* https://www.python.org/dev/peps/pep-0517 
* inspire from or integrate with tools like flit. May be a flit plugin ? https://flit.readthedocs.io/en/latest/pyproject_toml.html#the-pyproject-toml-config-file)

**Comment by henryiii** [comment]
scikit-build (classic) after #1185 is now a thin wrapper around scikit-build-core's setuptools plugin. Closing all open issues; if this still applies, please reopen or open an issue on scikit-build-core. Thanks!
